grails-app/services/uk/ac/ox/softeng/maurodatamapper/plugins/csv/datamodel/provider/importer/CsvDataModelImporterProviderService.groovy

```diff
@@ -72,73 +79,158 @@ class CsvDataModelImporterProviderService
+                    throw new ApiBadRequestException('CSV02',
+                                                     "Archive [${parameters.importFile.fileName}] containts directory [${zipEntry.name}] but should only contain CSV " +
+                                                     "files.")
+                } else if (zipEntry.getName().contains('/')) {
+                    throw new ApiBadRequestException('CSV03',
+                                                     "Archive [${parameters.importFile.fileName}] containts file with path [${zipEntry.name}] but should not contain " +
+                                                     "directories.")
+                }
+                File newFile = newFile(tempDir, zipEntry)
+                FileOutputStream fos = new FileOutputStream(newFile)
+                int len
+                while ((len = zis.read(buffer)) > 0) {
+                    fos.write(buffer, 0, len)
+                }
+                fos.close()
+
+                parametersList << parameters.clone().tap {it.importFile = new FileParameter(newFile.name, 'text/csv', newFile.bytes)}
+
+                zipEntry = zis.getNextEntry()
+            }
+            zis.closeEntry()
+            zis.close()
+        } else {
+            parametersList << parameters
+        }
+
+        def (Map<String, ColumnData> allColumns, Map<String, Map<String, ColumnData>> fileColumns) = getColumnDefinitions(parametersList, importFromArchive)
+
+        if (!importFromArchive) allColumns = fileColumns[parameters.importFile.fileName]
+        Map<String, DataType> allDataTypes = allColumns.collectEntries {[it.key, it.value.getDataType(dataTypes)]}
+
+        parametersList.each {csvParameters ->
+            String className
+            if (importFromArchive) className = csvParameters.importFile.fileName.take(csvParameters.importFile.fileName.lastIndexOf('.'))
+            else className = 'CSV fields'
+
+            DataClass fileClass = new DataClass(label: className)
+            dataModel.addToDataClasses(fileClass)
+
+            Map<String, ColumnData> columns = fileColumns[csvParameters.importFile.fileName]
+
+            Integer columnIndex = 0
+            columns.each {header, column ->
+                DataType dataType = allDataTypes[header]
+                if (dataType instanceof EnumerationType && !dataModel.findDataTypeByLabel(dataType.label)) {
+                    log.debug('Adding datatype with label {}', dataType.label)
+                    dataModel.addToDataTypes(dataType)
+                }
+                if (!dataType) log.warn('Unknown datatype {}', column.decideDataType())
+                DataElement dataElement = new DataElement(
+                    label: header,
+                    dataType: dataType,
+                    minMultiplicity: column.getMinMultiplicity(),
+                    maxMultiplicity: column.getMaxMultiplicity()).tap { setIndex(columnIndex)}
```

Comment:
rather than `tap` use the map field `idx` to set the index

Reply:
Done